flang/lib/Lower/Bridge.cpp

```diff
@@ -1240,6 +1262,43 @@ class FirConverter : public Fortran::lower::AbstractConverter {
                                        valueList, blockList);
   }
 
+  void genFIR(const Fortran::parser::AssociateConstruct &) {
+    for (auto &e : getEval().getNestedEvaluations())
+      genFIR(e);
+  }
+
+  void genFIR(const Fortran::parser::AssociateStmt &stmt) {
+    Fortran::lower::StatementContext stmtCtx;
+    localSymbols.pushScope();
+    for (auto &assoc :
+         std::get<std::list<Fortran::parser::Association>>(stmt.t)) {
+      auto &sym = *std::get<Fortran::parser::Name>(assoc.t).symbol;
+      const auto &rhsExpr =
+          *sym.get<Fortran::semantics::AssocEntityDetails>().expr();
+      if (!Fortran::evaluate::IsVariable(rhsExpr) ||
+          Fortran::evaluate::HasVectorSubscript(rhsExpr)) {
+        // Associating entity is assigned the value of the selector expression.
+        mapSymbolAttributes(Fortran::lower::pft::Variable(sym), stmtCtx);
+        genIntrinsicAssignment(ExprOfSym(sym), rhsExpr);
```

Comment:
This doesn't look correct to me. An association should be introducing a set of new bindings and not assignments to variables. Each associate name should bind to the value of selector. That could be a copy of the bridge box if the selector is a variable in the symbol table (SymMap) already or it may be a binding to the ssa-value result of lowering the selector expression. That approach will work regardless of the type of the expression and correctly alias reference objects, etc.

Reply:
The code is based on the distinctions in F18 Clause 19.5,1,6p3 -

_If the selector is a variable other than an array section having a vector subscript, the association is with the data object specified by the selector; otherwise, the association is with the value of the selector expression, which is evaluated prior to execution of the block._

I read this as two cases.

The case before the semicolon adds an alias to a host variable.  One special case is when the variable is a whole object.  In that case, the code can use the host variable mapping.  The more general subobject case is not done in this PR.

The case after the semicolon is implemented as an assignment to a variable that is placed in a symbol table scope for use during the associate statement construct.

Reply:
Right, I do not think your reading of the requirement/standard is wrong, but I think Eric's point is that something like simply calling `exv = genExprAddr(rhsExpr, stmtCtxt)`, and mapping the returned extended value should cover all cases. It would also avoid any extra work dealing with derived type/character/array complexity (see my answer to your other comments).

`genExprAddr` will create temps if the expression is not a variable, is a designator with a vector subscript (with an exception discussed below in [1]). Otherwise it will return an ExtendedValue describing the variable. Which is what you want here.

The only thing you would have to map the extended value to a symbol with something like:
```
genExprAddr(rhsExpr, stmtCtxt).match(
        [&](const fir::UnboxedValue value) {
           addSymbol(sym, value);
        },
        [&](const fir::CharBoxValue &ch) {
             addCharSymbol(sym,  ch.getAddr(), ch.getLen())
        },
        [&](const fir::ArrayBoxValue &arr) {
          if (arr.getLBounds().empty())
             addSymbolWithShape(sym, arr.getAddr(), arr.getExtents())
          else
            addSymbolWithBounds(sym,  arr.getAddr(), arr.getExtents(), arr.getLBounds())
        },

        [&](const fir::CharArrayBoxValue &arr) {
          if (arr.getLBounds().empty())
             addCharSymbolWithShape(sym,  arr.getAddr(), arr.getLen(), arr.getExtents())
          else
            addCharSymbolWithBounds(sym,  arr.getAddr(), arr.getLen(), arr.getExtents(), arr.getLBounds())
        },
        [&](const fir::BoxValue &arr) {
          TODO("derived types in associate construct");
        },
        [&](const fir::MutableBoxValue&) {
          fir::emitFatalError(loc, "genExprAddr should not return MutableBoxValue");
        });
```

Notes:

Other very important point, you will want to finalize the `stmtCtxt` used in genExprAddr when leaving the associate construct, but not before, that will deallocate any array temps that may have been created for array selector like (x => a + b).

One drawback of using genExprAddr is that you would currently not control the temp name, it would be a compiler generated temp name. I think it is not a functional issue currently. Someone could add a way to name the main temp created by `genExprAddr()` if needed.

[1] The exception is that `genExprAddr` is currently creating temps for all designator that are array sections, whether they have a vector subscript or not. One of the reason behind that is that ExtendedValues currently do not allow representing such variable. The problem already is visible in call lowering, when we are passing a temp when we should not. Eric latest PR #580 is addressing part of this problem by creating a different expression lowering entry point, `createSomeArrayBox`, that will create a `fir.box` mlir value describing the variable.
I say part of the issue because you would still need to detect that you are in this case (i.e, array variable with array sections other than vector subscript) and have a bit of work to map the `fir.box` to a symbol (You would probably need to read the fir.box, and that may not be the best approach). That's probably an OK TODO to leave at first.

Reply:
Right, Jean. That's the idea. We don't necessarily have to lower expressions into temps here, but it should be straightforward to use that code. There is probably a bit of work to lower symbol boxes (in the bridge) to FIR boxes (via embox) in the appropriate cases. I'm not sure there are ready interfaces to the existing code presently (or in other ongoing PRs).